Themis/ViewModels/CorrectionSidebar/ProblemStatementCellViewModel.swift

```diff
@@ -32,47 +32,91 @@ class ProblemStatementCellViewModel: ObservableObject {
+
         while index < problemStatement.endIndex {
             var substring: String
 
             guard let rangeStart = problemStatement.range(of: "@startuml", range: index..<problemStatement.endIndex), let rangeEnd = problemStatement.range(of: "@enduml", range: index..<problemStatement.endIndex) else {
                 substring = String(problemStatement[index...])
-                substring = deleteTestCases(substring)
+                substring = replaceProblemStatementTestCases(substring, feedbacks)
                 problemStatementParts.append(ProblemStatementMD(text: substring))
                 return
             }
 
             // Append markdown
             substring = String(problemStatement[index..<rangeStart.lowerBound])
-            substring = deleteTestCases(substring)
+            substring = replaceProblemStatementTestCases(substring, feedbacks)
             problemStatementParts.append(ProblemStatementMD(text: substring))
 
             // Append plantuml
             substring = String(problemStatement[rangeStart.lowerBound...rangeEnd.upperBound])
-            substring = replaceTestsColor(substring, colorScheme)
+            substring = replaceUMLTestsColor(substring, feedbacks, colorScheme)
             problemStatementParts.append(ProblemStatementPlantUML(colorScheme: colorScheme, text: substring))
 
             index = problemStatement.index(rangeEnd.upperBound, offsetBy: 1)
         }
     }
 
-    func deleteTestCases(_ problemStatement: String) -> String {
-        return problemStatement
-            .replacingOccurrences(of: "[task][", with: "") // to remove symbol
-            .replacingOccurrences(of: "](.*())", with: "", options: .regularExpression) // to remove tests
+    private func replaceProblemStatementTestCases(_ problemStatement: String, _ feedbacks: [AssessmentFeedback]) -> String {
+        var problemStatementText = problemStatement
+
+        for feedback in feedbacks {
+            if let feedbackText = feedback.text {
+                problemStatementText = problemStatementText
+                    .replacingOccurrences(of: "\(feedbackText)", with: feedback.credits == 0.0 ? "0" : "1")
+            }
+        }
+
+        let lines = problemStatementText.components(separatedBy: "\n")
+        problemStatementText = ""
+
+        for var line in lines {
+            if line.contains("[task][") && line.contains("0") {
+                line = calculateScoreOfPassingTests(line)
+                line = line.replacingOccurrences(of: "[task][", with: " ![Test failed](asset:///TestFailedSymbol) **")
+            } else if line.contains("[task][") && line.contains("1") {
+                line = calculateScoreOfPassingTests(line)
+                line = line.replacingOccurrences(of: "[task][", with: " ![Test passed](asset:///TestPassedSymbol) **")
+            }
+
+            problemStatementText.append(line + "\n")
+        }
+
+        return problemStatementText
```

Comment:
Just an idea (doesn't have to be implemented): This looks a lot like business logic, maybe it should be in a ViewModel?

Reply:
It is in a ViewModel (ProblemStatementCellViewModel).